util/db/repository.go

```diff
@@ -40,517 +36,291 @@ const (
 	githubAppPrivateKey = "githubAppPrivateKey"
 )
 
+// repositoryBackend defines the API for types that wish to provide interaction with repository storage
+type repositoryBackend interface {
+	CreateRepository(ctx context.Context, r *appsv1.Repository) (*appsv1.Repository, error)
+	GetRepository(ctx context.Context, repoURL string) (*appsv1.Repository, error)
+	ListRepositories(ctx context.Context, repoType *string) ([]*appsv1.Repository, error)
+	UpdateRepository(ctx context.Context, r *appsv1.Repository) (*appsv1.Repository, error)
+	DeleteRepository(ctx context.Context, repoURL string) error
+	RepositoryExists(ctx context.Context, repoURL string) (bool, error)
+
+	CreateRepoCreds(ctx context.Context, r *appsv1.RepoCreds) (*appsv1.RepoCreds, error)
+	GetRepoCreds(ctx context.Context, repoURL string) (*appsv1.RepoCreds, error)
+	ListRepoCreds(ctx context.Context) ([]string, error)
+	UpdateRepoCreds(ctx context.Context, r *appsv1.RepoCreds) (*appsv1.RepoCreds, error)
+	DeleteRepoCreds(ctx context.Context, name string) error
+	RepoCredsExists(ctx context.Context, repoURL string) (bool, error)
+
+	GetAllHelmRepoCreds(ctx context.Context) ([]*appsv1.RepoCreds, error)
+}
+
 func (db *db) CreateRepository(ctx context.Context, r *appsv1.Repository) (*appsv1.Repository, error) {
-	repos, err := db.settingsMgr.GetRepositories()
+	secretBackend := repositoryBackend(&secretsRepositoryBackend{db: db})
```

Comment:
Nitpick all you want. Thats what reviews are for and I am thankful for every remark 😉.

Would you also cache the instances? Creating a new instance with every call seems okay in my eyes, since the struct is rather small. Caching the instances on the other hand will make creating new `db` instances more complicated on the other hand,as well as create a weird chain of dependencies, since `db` would need the repository backends and the repository backend would need an instance of `db`.

I created a factory function for now as you suggested, but passed on the instance caching for now. Is that fine with you?

Reply:
The factory function is totally fine. As long as we won't have to change code in multiple places in the future (if e.g. one of backends creation require some other dependency)

Reply:
Nice, so we are good on that one!